File: android/pi-agent-core/src/main/kotlin/com/pimobile/agent/AgentLoop.kt
package com.pimobile.agent

import com.pimobile.ai.*
import kotlinx.coroutines.flow.*
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject

class AgentLoop(private val llmClient: LlmClient) {

    private val json = Json { ignoreUnknownKeys = true }

    fun run(
        model: ModelDefinition,
        context: Context,
        tools: List<Tool>,
        apiKey: String
    ): Flow<AgentEvent> = flow {
        val toolDefs = tools.map {
            ToolDefinition(
                name = it.name,
                description = it.description,
                parameters = it.parametersSchema
            )
        }
        val ctxWithTools = context.copy(tools = toolDefs)

        var loopCount = 0
        val maxLoops = 50

        while (loopCount < maxLoops) {
            loopCount++

            val response = collectStreamedResponse(model, ctxWithTools, apiKey)

            emit(AgentEvent.AssistantMessage(
                content = response.textContent,
                thinking = response.thinking
            ))

            if (response.toolCalls.isNullOrEmpty()) {
                emit(AgentEvent.Done)
                break
            }

            val results = mutableListOf<ToolResult>()
            for (call in response.toolCalls) {
                emit(AgentEvent.ToolCallStarted(call.name, call.input))
                val result = executeTool(call, tools)
                results.add(ToolResult(call.id, result.output, result.isError))
                emit(AgentEvent.ToolCallCompleted(call.name, result))
            }

            ctxWithTools.messages.add(Message(
                role = Role.ASSISTANT,
                content = MessageContent.Text(response.textContent),
                toolCalls = response.toolCalls,
                thinking = response.thinking
            ))
            ctxWithTools.messages.add(Message(
                role = Role.USER,
                content = MessageContent.Text(""),
                toolResults = results
            ))
        }

        if (loopCount >= maxLoops) {
            emit(AgentEvent.Error("Agent loop exceeded maximum iterations ($maxLoops)"))
            emit(AgentEvent.Done)
        }
    }

    private suspend fun FlowCollector<AgentEvent>.collectStreamedResponse(
        model: ModelDefinition,
        context: Context,
        apiKey: String
    ): CollectedResponse {
        val textBuilder = StringBuilder()
        val thinkingBuilder = StringBuilder()
        val toolCalls = mutableMapOf<String, ToolCallAccumulator>()
        var hasThinking = false

        llmClient.stream(model, context, apiKey).collect { event ->
            when (event) {
                is StreamEvent.TextDelta -> {
                    textBuilder.append(event.delta)
                    emit(AgentEvent.StreamDelta(event.delta))
                }
                is StreamEvent.ThinkingDelta -> {
                    thinkingBuilder.append(event.delta)
                    hasThinking = true
                    emit(AgentEvent.ThinkingDelta(event.delta))
                }
                is StreamEvent.ToolCallStart -> {
                    toolCalls[event.id] = ToolCallAccumulator(event.id, event.name, StringBuilder())
                }
                is StreamEvent.ToolCallDelta -> {
                    toolCalls[event.id]?.arguments?.append(event.argumentsDelta)
                }
                is StreamEvent.ToolCallEnd -> {
                    // Tool call finalized
                }
                is StreamEvent.Usage -> {
                    emit(AgentEvent.UsageUpdate(event.inputTokens, event.outputTokens))
                }
                is StreamEvent.Error -> {
                    emit(AgentEvent.Error(event.message))
                }
                is StreamEvent.Done -> {
                    // Stream complete
                }
            }
        }

        val completedToolCalls = if (toolCalls.isNotEmpty()) {
            toolCalls.values.map { acc ->
                ToolCall(
                    id = acc.id,
                    name = acc.name,
                    arguments = acc.arguments.toString().ifBlank { "{}" }
                )
            }
        } else null

        return CollectedResponse(
            textContent = textBuilder.toString(),
            thinking = if (hasThinking) thinkingBuilder.toString() else null,
            toolCalls = completedToolCalls
        )
    }

    private suspend fun executeTool(call: ToolCall, tools: List<Tool>): AgentToolResult {
        val tool = tools.find { it.name == call.name }
            ?: return AgentToolResult(call.id, "Error: Unknown tool '${call.name}'", isError = true)

        val validation = SchemaValidator.validate(tool.parametersSchema, call.input)
        if (!validation.isValid) {
            return AgentToolResult(
                call.id,
                "Validation error: ${validation.errors.joinToString()}",
                isError = true
            )
        }

        return try {
            tool.execute(call.input).copy(toolCallId = call.id)
        } catch (e: Exception) {
            AgentToolResult(
                call.id,
                "Tool error: ${e.message}",
                ToolResultDetails.Error(e.message ?: "", null),
                true
            )
        }
    }
}

private data class ToolCallAccumulator(
    val id: String,
    val name: String,
    val arguments: StringBuilder
)

private data class CollectedResponse(
    val textContent: String,
    val thinking: String?,
    val toolCalls: List<ToolCall>?
)

File: ios/PiAgentCore/Sources/PiAgentCore/AgentLoop.swift
import Foundation
import PiAI
import os

private let logger = Logger(subsystem: "com.pi.ai", category: "AgentLoop")

public final class AgentLoop: Sendable {
    private let llmClient: LlmClient
    private let validator: SchemaValidator

    public init(llmClient: LlmClient) {
        self.llmClient = llmClient
        self.validator = SchemaValidator()
    }

    /// Run the agent loop: stream a response, execute tool calls, repeat until done.
    public func run(
        model: ModelDefinition,
        context: inout Context,
        tools: [any Tool],
        apiKey: String,
        maxIterations: Int = 20
    ) -> AsyncThrowingStream<AgentEvent, Error> {
        // Capture context as a let copy; mutability moves inside the Task.
        var localContext = context
        localContext.tools = tools.map { tool in
            ToolDefinition(name: tool.name, description: tool.description, parameters: tool.parametersSchema)
        }
        let capturedContext = localContext
        let client = self.llmClient
        let validator = self.validator

        return AsyncThrowingStream { continuation in
            Task {
                do {
                    var localContext = capturedContext
                    var iteration = 0

                    while iteration < maxIterations {
                        iteration += 1
                        logger.debug("[AgentLoop] Iteration \(iteration)/\(maxIterations)")

                        // Stream response from LLM
                        var fullText = ""
                        var fullThinking = ""
                        var pendingToolCalls: [ToolCall] = []
                        var currentToolCallArgs: [String: String] = [:]

                        let stream = client.stream(model: model, context: localContext, apiKey: apiKey)

                        for try await event in stream {
                            switch event {
                            case .textDelta(let delta):
                                fullText += delta
                                continuation.yield(.streamDelta(delta))

                            case .thinkingDelta(let delta):
                                fullThinking += delta
                                continuation.yield(.thinkingDelta(delta))

                            case .toolCallStart(let id, let name):
                                currentToolCallArgs[id] = ""
                                logger.info("[AgentLoop] Tool call started: \(name) (id=\(id))")
                                continuation.yield(.toolCallStarted(name: name, input: .null))
                                pendingToolCalls.append(ToolCall(id: id, name: name, arguments: ""))

                            case .toolCallDelta(let id, let argsDelta):
                                currentToolCallArgs[id, default: ""] += argsDelta

                            case .toolCallEnd(let id):
                                if let index = pendingToolCalls.firstIndex(where: { $0.id == id }) {
                                    let args = currentToolCallArgs[id] ?? "{}"
                                    pendingToolCalls[index] = ToolCall(
                                        id: pendingToolCalls[index].id,
                                        name: pendingToolCalls[index].name,
                                        arguments: args
                                    )
                                    logger.info("[AgentLoop] Tool call complete: \(pendingToolCalls[index].name) args=\(args.prefix(500))")
                                }

                            case .usage(let input, let output):
                                continuation.yield(.usageUpdate(inputTokens: input, outputTokens: output))

                            case .done:
                                break

                            case .error(let message):
                                logger.error("[AgentLoop] Stream error: \(message)")
                                continuation.yield(.error(message))
                            }
                        }

                        logger.info("[AgentLoop] LLM response: \(fullText.count) chars, \(pendingToolCalls.count) tool call(s)")
                        if !fullText.isEmpty {
                            logger.debug("[AgentLoop] Response text: \(fullText.prefix(500))")
                        }

                        // Emit full assistant message
                        let thinking: String? = fullThinking.isEmpty ? nil : fullThinking
                        continuation.yield(.assistantMessage(content: fullText, thinking: thinking))

                        // If no tool calls, we're done
                        if pendingToolCalls.isEmpty {
                            localContext.messages.append(Message(
                                role: .assistant,
                                content: .text(fullText),
                                thinking: thinking
                            ))
                            break
                        }

                        // Execute tool calls
                        logger.info("[AgentLoop] Executing \(pendingToolCalls.count) tool call(s)")
                        var toolResults: [ToolResult] = []
                        for toolCall in pendingToolCalls {
                            let result = await executeTool(toolCall, tools: tools, validator: validator)
                            logger.info("[AgentLoop] Tool result [\(toolCall.name)]: isError=\(result.isError) output=\(result.output.prefix(300))")
                            continuation.yield(.toolCallCompleted(name: toolCall.name, result: result))
                            toolResults.append(ToolResult(
                                toolCallId: result.toolCallId,
                                output: result.output,
                                isError: result.isError
                            ))
                        }

                        // Append assistant message + tool results to context
                        localContext.messages.append(Message(
                            role: .assistant,
                            content: .text(fullText),
                            toolCalls: pendingToolCalls,
                            thinking: thinking
                        ))
                        localContext.messages.append(Message(
                            role: .user,
                            content: .text(""),
                            toolResults: toolResults
                        ))
                    }

                    if iteration >= maxIterations {
                        logger.warning("[AgentLoop] Hit max iterations (\(maxIterations))")
                    }

                    continuation.yield(.done)
                    continuation.finish()
                } catch {
                    logger.error("[AgentLoop] Error: \(error.localizedDescription)")
                    continuation.yield(.error(error.localizedDescription))
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    private func executeTool(_ call: ToolCall, tools: [any Tool], validator: SchemaValidator) async -> AgentToolResult {
        guard let tool = tools.first(where: { $0.name == call.name }) else {
            return AgentToolResult(
                toolCallId: call.id,
                output: "Error: Unknown tool '\(call.name)'",
                details: .error(message: "Unknown tool", code: "UNKNOWN_TOOL"),
                isError: true
            )
        }

        // Parse arguments
        let input = call.input

        // Validate against schema
        let validation = validator.validate(input, against: tool.parametersSchema)
        if !validation.isValid {
            let errorMessages = validation.errors.map { $0.localizedDescription }.joined(separator: "; ")
            return AgentToolResult(
                toolCallId: call.id,
                output: "Validation error: \(errorMessages)",
                details: .error(message: errorMessages, code: "VALIDATION_ERROR"),
                isError: true
            )
        }

        do {
            let result = try await tool.execute(input: input)
            return AgentToolResult(
                toolCallId: call.id,
                output: result.output,
                details: result.details,
                isError: result.isError
            )
        } catch {
            logger.error("[AgentLoop] Tool '\(call.name)' error: \(error.localizedDescription)")
            return AgentToolResult(
                toolCallId: call.id,
                output: "Error executing tool '\(call.name)': \(error.localizedDescription)",
                details: .error(message: error.localizedDescription, code: "EXECUTION_ERROR"),
                isError: true
            )
        }
    }
}
